pkg/codegen/importer/hcl2.go

```diff
@@ -365,11 +366,109 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
 	return generateValue(property.Type, value)
 }
 
+// reduceUnionType reduces the given union type to a simpler type that potentially matches the value.
+// When the value type is primitive, choose the first element type of the union elements that is of the same type.
+// When the value is an object, use the discriminator to choose the element type.
+func reduceUnionType(schemaUnion *schema.UnionType, value resource.PropertyValue) schema.Type {
+	switch {
+	case value.IsObject():
+		// If the value is an object, use the discriminator to choose the element type.
+		if schemaUnion.Discriminator == "" {
+			// If there is no discriminator, return first type which is an object
```

Comment:
I don't agree with this. If there is no discriminator, we need to return the first type it is legal to assign `value` to. This is a structural type check.

For example, given the value `{ foo: 1 }` and `Union[{ foo: string }, { bar: Union[none, string], foo: number }` we should resolve to the second element, not the first.

An example of where this would be necessary is 
https://github.com/pulumi/pulumi-aws-native/blob/edc3d04caddc8996061d1ce2a8b80fbe5236b313/provider/cmd/pulumi-resource-aws-native/schema.json#L13430-L13451.

Reply:
I've implemented the structural type check to find the best fit + tests ✅